AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -1491,13 +1469,116 @@ - (void)measureNodeWithBoundsIfNecessary:(CGRect)bounds
+  auto pendingLayout = [=]() -> std::shared_ptr<ASDisplayNodeLayout> {
+    // Check if  the pending display node layout can be used to transition to
+    if (_pendingDisplayNodeLayout != nullptr &&
+        _pendingDisplayNodeLayout->isDirty() == NO &&
+        CGSizeEqualToSize(_pendingDisplayNodeLayout->layout.size, ASCeilSizeValues(bounds.size))) {
+      // We assume the _pendingDisplayNodeLayout was created by layoutThatFits: to set the size of the node
+      // now it's time to apply it and to become the _calculatedDisplayNodeLayout
+      return _pendingDisplayNodeLayout;
+    }
+  
+    // The _pendingDisplayNodeLayout can not be used to be applied we need to figure out we need to calculate a new
+    // layout based on a constrainedSize
+  
+    // Use as default constrained size the bounds
+    CGSize parentSize = bounds.size;
+    ASSizeRange constrainedSize = [self constrainedSizeForLayoutPass];
+    
+    return std::make_shared<ASDisplayNodeLayout>(
+      [self calculateLayoutThatFits:constrainedSize restrictedToSize:self.style.size relativeToParentSize:parentSize],
+      constrainedSize,
+      parentSize
+    );
+  }();
+  
+  // If the size of the new layout to apply did change from the current bounds, invalidate the whole tree up
+  // so the root node can handle a resizing if necessary
+  if (CGSizeEqualToSize(ASCeilSizeValues(bounds.size), pendingLayout->layout.size) == NO) {
+    [self invalidateSize];
+  }
+
+  if (didCreateNewContext) {
+    ASLayoutElementClearCurrentContext();
+  } else if (didOverrideExistingContext) {
+    context.needsVisualizeNode = !context.needsVisualizeNode;
+    ASLayoutElementSetCurrentContext(context);
+  }
+
+  // Finally transition to pendingLayout
+  ASDisplayNodeAssertNotNil(pendingLayout->layout, @"pendintLayout->layout should not be nil! %@", self);
+
+  _pendingLayoutTransition = [[ASLayoutTransition alloc] initWithNode:self
+                                                        pendingLayout:pendingLayout
+                                                       previousLayout:previousLayout];
+
+  // Only complete the pending layout transition if the node is not a subnode of a node that is currently
+  // in a layout transition
+  if (ASHierarchyStateIncludesLayoutPending(_hierarchyState) == NO) {
+    // Complete the pending layout transition immediately
+    [self _completePendingLayoutTransition];
+  }
+}
+
+- (ASSizeRange)constrainedSizeForLayoutPass
+{
+  // Use as default constrained size the bounds
+  CGRect bounds = self.threadSafeBounds;
+  ASSizeRange constrainedSize = ASSizeRangeMake(bounds.size);
+  
+  // Checkout if constrained size of pending or calculated display node layout can be used
+  if (_pendingDisplayNodeLayout != nullptr && CGSizeEqualToSize(_pendingDisplayNodeLayout->layout.size, ASCeilSizeValues(bounds.size))) {
+    // We assume the size from the last returned layoutThatFits: layout was applied so use the pejnding display node
```

Comment:
typo in word "pending"

Can shorten this comment to one line without being the longest -- "so use the pending layout constrained size"